relay-server/src/services/projects/project/info.rs

```diff
@@ -133,6 +134,30 @@ impl ProjectInfo {
             return Err(DiscardReason::FeatureDisabled(*disabled_feature));
         }
 
+        if !envelope.meta().is_from_internal_relay()
+            && self
+                .config()
+                .trusted_relay_settings
+                .verify_signature
+                .is_verify_timestamp()
+        {
+            match envelope.meta().signature() {
+                Some(RelaySignature::Valid(signature)) => {
+                    if !signature.verify_any(
+                        &self.config.trusted_relays,
+                        envelope.received_at(),
+                        Some(Duration::seconds(config.signature_max_age() as i64)),
+                    ) {
+                        return Err(DiscardReason::InvalidSignature);
+                    }
+                }
+                Some(RelaySignature::Invalid(_)) => {
+                    return Err(DiscardReason::InvalidSignature);
+                }
+                None => return Err(DiscardReason::MissingSignature),
+            }
+        }
```

Comment:
I think this is a bit better in its own function, then you can exit early on each condition and match on the verification, which has the upside if we add another method the compiler will tell us that we have to do something.